src/test/ui/dylibs/diamonds_core.rs

```diff
@@ -0,0 +1,21 @@
+// ignore-test Not a test. Used by the other tests.
+
+pub fn sanity_check() {
+    // Sanity-check: getting back non-trival values from addr functions
+    assert_ne!(z::s_addr(), z::t_addr());
+    assert_ne!(z::s_m_addr(), z::s_m_i_addr());
+    assert_ne!(z::s_m_i_addr(), z::s_m_j_addr());
+    assert_ne!(z::s_m_i_addr(), z::s_m_i_a_addr());
+}
+
+pub fn check_linked_function_equivalence() {
+    // Check that the linked functions are the same code by comparing their
+    // underlying addresses.
```

Comment:
I wouldn't be surprised if the dynamic linker would resolve each symbol with the same name to the same address at runtime even if it is duplicated between dylibs.

Reply:
Possibly... I have certainly been thinking that the testing methodology here might be flawed. I'm still musing about the best way to write these tests (i.e. to detect unwanted duplication of static libraries across dynamic libraries being loaded).

I'm still planning to invest some effort into a `run-make` test, to check the most basic scenario of deleting a dynamic library and [observing that this injects](https://github.com/rust-lang/rust/pull/88101#issuecomment-900483063) a failure to run the binary. But that isn't quite the same check that I'm trying to put in here.

Reply:
You could do an objdump of dylibs to check which symbols are defined and imported. Imported symbols use 0 as specified address.

Reply:
Okay I added a run-make test in 21778cba8e73fa28849ad31d0839df3b413947fe; it does two things:

 1. it uses the pre-existing pattern of building, running to success, then deleting dylibs, and checking that a re-run attempt fails. (I think this testing methodology is fine, as far as it goes.)
 2. it introduces a new pattern of having methods *in* the compiled code that differ depending on whether one is linked to the rlib or to the dylib. This is controlled via `cfg` flags. This way, we can directly determine which object code has been linked and loaded when we finally run the program.